packages/hashtower.circom/circuits/hashtower-hash-chain.circom

```diff
@@ -0,0 +1,228 @@
+
+    signal leadingOnes[N] <== LeadingOnes(N)(prefixLen);
+    signal isGood[N];
+    var goodCount = 0;
+    for (var i = 0; i < N; i++) {
+        isGood[i] <== AND()(IsEqual()([in[i], v]), leadingOnes[i]);
+        goodCount += isGood[i];
+    }
+    out <== IsNonZero()(goodCount);
+}
+
+// Compuates a Merkle root at rootLv made from childrens[0 .. rootLv - 1][] and leaf.
+//
+// Each childrens[i] must include the digest of childrens[i - 1] for i = 1 ... rootLv - 1
+// childrens[0] must include leaf.
+//
+// root = digest of childrens[rootLv - 1] if rootLv > 0
+// root = leaf if rootLv == 0
+template CheckMerkleProofAndComputeRoot(H, W) {
+    signal input childrens[H - 1][W]; // we only check childrens[] *below* rootLv
+    signal input rootLv; // 0 <= rootLv <= H - 1  (otherwise will be rejected)
+    signal input leaf;
+    signal output root;
+
+    // TBI: to be included
+    //
+    // childrens[lv] must include TBI[lv]  for all lv in [0 ... rootLv - 1]
+    //
+    // TBI[0] = leaf
+    // TBI[lv] = digest of children[lv - 1]  for all lv in [1 ... rootLv]
+    // TBI[rootLv] is the root
+    //
+    // For rootLv = 3, H = 5:                                                         mustInclude[]
+    //                                  TBI[4]                                        0
+    // rootLv =>                        TBI[3] <== digest of childrens[2] ==> *root*  0
+    //           childrens[2]  include  TBI[2] <== digest of childrens[1]             1
+    //           childrens[1]  include  TBI[1] <== digest of childrens[0]             1
+    //           childrens[0]  include  TBI[0] <== *leaf*                             1
+    signal TBI[H];
+    TBI[0] <== leaf;
+
+    // LeadingOnes() will also help us enforce rootLv <= H - 1
+    signal mustInclude[H - 1] <== LeadingOnes(H - 1)(rootLv);
+    for (var lv = 0; lv < H - 1; lv++) {
+        // Either  childrens[lv] include TBI[lv]   or   !mustInclude[lv]
+        Must()(OR()(
+            Include(W)( childrens[lv], TBI[lv] ),
+            NOT()( mustInclude[lv] )
+        ));
+        TBI[lv + 1] <== HashChain(W)(childrens[lv], W);
+    }
+    root <== PickOne(H)(TBI, rootLv); // root = TBI[rootLv]
+}
+
+function computeDataHeight(levelLengths, bitsPerLevel) {
+    var h = 0;
+    while (levelLengths > 0) {
+        h++;
+        levelLengths >>= bitsPerLevel;
+    }
+    return h;
+}
+function computeSingleLevelLength(levelLengths, lv, bitsPerLevel) {
+    var mask = (1 << bitsPerLevel) - 1;
+    return (levelLengths >> (lv * bitsPerLevel)) & mask;
+}
+template ComputeDataHeightAndLevelLengthArray(H, W, bitsPerLevel) {
+    signal input levelLengths;
+    signal output dataHeight;
+    signal output levelLengthArray[H];
+
+    // compute
+    dataHeight <-- computeDataHeight(levelLengths, bitsPerLevel);
+    for (var lv = 0; lv < H; lv++) {
+        levelLengthArray[lv] <-- computeSingleLevelLength(levelLengths, lv, bitsPerLevel);
```

Comment:
The code relying on the following statement:

If
1. levelLengthArray[i] belongs to [1, W] for i in [0, dataHeight - 1], otherwise levelLengthArray[i] = 0
2. dataHeight belongs to [0, H - 1]
3. the W^i weighted sum of levelLengthArray = levelLengths

then levelLengthArray[] and dataHeight will be unique.

A very rough analogy would be like in the 10-base system, the only sequence matches with `492` could only be the length 3 `[2, 9, 4]` .

Reply:
Thank you for clarifying. I think this makes sense.

On line 188 the `LessEqThan` operation needs to have it's input `levelLengthArray[lv]` constrained to `bitsPerLevel`. You can do this by

```
Num2Bits(bitsPerLevel)([levelLengthArray[lv]]);
```